src/tile-component.coffee

```diff
@@ -0,0 +1,285 @@
+        newLineIds.push(id)
+        newLinesHTML += @buildLineHTML(id)
+        @screenRowsByLineId[id] = lineState.screenRow
+        @lineIdsByScreenRow[lineState.screenRow] = id
+        @oldTileState.lines[id] = cloneObject(lineState)
+
+    return unless newLineIds?
+
+    WrapperDiv.innerHTML = newLinesHTML
+    newLineNodes = _.toArray(WrapperDiv.children)
+    for id, i in newLineIds
+      lineNode = newLineNodes[i]
+      @lineNodesByLineId[id] = lineNode
+      @domNode.appendChild(lineNode)
+
+    return
+
+  buildLineHTML: (id) ->
+    {scrollWidth} = @newState
+    {screenRow, tokens, text, top, lineEnding, fold, isSoftWrapped, indentLevel, decorationClasses} = @newTileState.lines[id]
+
+    classes = ''
+    if decorationClasses?
+      for decorationClass in decorationClasses
+        classes += decorationClass + ' '
+    classes += 'line'
+
+    lineHTML = "<div class=\"#{classes}\" style=\"position: absolute; top: #{top}px; width: #{scrollWidth}px;\" data-screen-row=\"#{screenRow}\">"
+
+    if text is ""
+      lineHTML += @buildEmptyLineInnerHTML(id)
+    else
+      lineHTML += @buildLineInnerHTML(id)
+
+    lineHTML += '<span class="fold-marker"></span>' if fold
+    lineHTML += "</div>"
+    lineHTML
+
+  buildEmptyLineInnerHTML: (id) ->
+    {indentGuidesVisible} = @newState
+    {indentLevel, tabLength, endOfLineInvisibles} = @newTileState.lines[id]
+
+    if indentGuidesVisible and indentLevel > 0
+      invisibleIndex = 0
+      lineHTML = ''
+      for i in [0...indentLevel]
+        lineHTML += "<span class='indent-guide'>"
+        for j in [0...tabLength]
+          if invisible = endOfLineInvisibles?[invisibleIndex++]
+            lineHTML += "<span class='invisible-character'>#{invisible}</span>"
+          else
+            lineHTML += ' '
+        lineHTML += "</span>"
+
+      while invisibleIndex < endOfLineInvisibles?.length
+        lineHTML += "<span class='invisible-character'>#{endOfLineInvisibles[invisibleIndex++]}</span>"
+
+      lineHTML
+    else
+      @buildEndOfLineHTML(id) or '&nbsp;'
+
+  buildLineInnerHTML: (id) ->
+    {indentGuidesVisible} = @newState
+    {tokens, text, isOnlyWhitespace} = @newTileState.lines[id]
+    innerHTML = ""
+
+    scopeStack = []
+    for token in tokens
+      innerHTML += @updateScopeStack(scopeStack, token.scopes)
+      hasIndentGuide = indentGuidesVisible and (token.hasLeadingWhitespace() or (token.hasTrailingWhitespace() and isOnlyWhitespace))
+      innerHTML += token.getValueAsHtml({hasIndentGuide})
+
+    innerHTML += @popScope(scopeStack) while scopeStack.length > 0
+    innerHTML += @buildEndOfLineHTML(id)
+    innerHTML
+
+  buildEndOfLineHTML: (id) ->
+    {endOfLineInvisibles} = @newTileState.lines[id]
+
+    html = ''
+    if endOfLineInvisibles?
+      for invisible in endOfLineInvisibles
+        html += "<span class='invisible-character'>#{invisible}</span>"
+    html
+
+  updateScopeStack: (scopeStack, desiredScopeDescriptor) ->
```

Comment:
You're going to need to integrate my changes on https://github.com/atom/atom/pull/6757 unfortunately but it shouldn't be a huge deal.